round-based/src/mpc/mod.rs

```diff
@@ -0,0 +1,329 @@
+//! Party of MPC protocol
+//!
+//! [`MpcParty`] is party of MPC protocol, connected to network, ready to start carrying out the protocol.
+//!
+//! ```rust
+//! use round_based::{Incoming, Outgoing};
+//!
+//! # #[derive(round_based::ProtocolMsg)]
+//! # enum KeygenMsg {}
+//! # struct KeyShare;
+//! # struct Error;
+//! # type Result<T> = std::result::Result<T, Error>;
+//! # async fn doc() -> Result<()> {
+//! async fn keygen<M>(party: M, i: u16, n: u16) -> Result<KeyShare>
+//! where
+//!     M: round_based::Mpc<Msg = KeygenMsg>
+//! {
+//!     // ...
+//! # unimplemented!()
+//! }
+//! async fn connect() ->
+//!     impl futures::Stream<Item = Result<Incoming<KeygenMsg>>>
+//!         + futures::Sink<Outgoing<KeygenMsg>, Error = Error>
+//!         + Unpin
+//! {
+//!     // ...
+//! # round_based::_docs::fake_delivery()
+//! }
+//!
+//! let delivery = connect().await;
+//! let party = round_based::mpc::connected(delivery);
+//!
+//! # let (i, n) = (1, 3);
+//! let keyshare = keygen(party, i, n).await?;
+//! # Ok(()) }
+//! ```
+
+use crate::{
+    round::{RoundInfo, RoundStore},
+    Outgoing, PartyIndex,
+};
+
+pub mod party;
+
+#[doc(no_inline)]
+pub use self::party::{Halves, MpcParty};
+
+/// Abstracts functionalities needed for MPC protocol execution
+pub trait Mpc {
+    /// Protocol message
+    type Msg;
+
+    /// Returned in [`Self::finish_setup`]
+    type Exec: MpcExecution<Msg = Self::Msg, SendErr = Self::SendErr>;
+    /// Error indicating that sending a message has failed
+    type SendErr;
+
+    /// Registers a round
+    fn add_round<R>(&mut self, round: R) -> <Self::Exec as MpcExecution>::Round<R>
+    where
+        R: RoundStore,
+        Self::Msg: RoundMsg<R::Msg>;
+
+    /// Completes network setup
+    ///
+    /// Once this method is called, no more rounds can be added,
+    /// but the protocol can receive and send messages.
+    fn finish_setup(self) -> Self::Exec;
+}
+
+/// Abstracts functionalities needed for MPC protocol execution
+pub trait MpcExecution {
+    /// Witness that round was registered
+    ///
+    /// It is used to retrieve messages in [`MpcExecution::complete`].
+    type Round<R: RoundInfo>;
+
+    /// Protocol message
+    type Msg;
+
+    /// Error indicating that completing a round has failed
+    type CompleteRoundErr<E>;
+    /// Error indicating that sending a message has failed
+    type SendErr;
+
+    /// Returned by [`.send_many()`](Self::send_many)
+    type SendMany: SendMany<Exec = Self, Msg = Self::Msg, SendErr = Self::SendErr>;
+
+    /// Completes the round
+    ///
+    /// Waits until all messages in the round `R` are received, returns the received messages.
+    async fn complete<R>(
+        &mut self,
+        round: Self::Round<R>,
+    ) -> Result<R::Output, Self::CompleteRoundErr<R::Error>>
```

Comment:
This phrasing implies that it waits until all _other_ parties receive messages, which I'm pretty sure it doesn't do for our impls. Would rephrase the second line of docs with something like:
> Waits until we receive all the messages in the round `R` from other parties in the protocol and returns those messages

Reply:
Ah yes, that's better wording